File: src/alpaca_agents/agent_desk.py
"""Read-only Agent Desk projection. No controller, broker, LLM or rule execution.

Cycles are recorded summaries, not a live event stream. Exact decision-key joins
may add a journaled idea and its *current* journal status; neither is an approval
capability. Never infer an entry/exit relationship from a ticker or timestamp.
"""
from collections import Counter
from datetime import datetime, timezone
from decimal import Decimal, DecimalException
import hashlib
import json
from pathlib import Path
import re
import sqlite3
from urllib.parse import parse_qs, urlsplit

from .dashboard_chat import redact_secrets

MAX_BYTES = 512 * 1024
MAX_CYCLES = 20
MAX_ROWS = 12
MAX_LINE = 64 * 1024
ID = re.compile(r"(?:[a-f0-9]{32}|legacy-[a-f0-9]{24})\Z")
ROLES = (
    ("spotter", "Star / Setups", "Spotter"),
    ("prior", "Research / Evidence", "Prior"),
    ("edge", "Research / Validation gaps", "Edge"),
    ("risk", "Moon / Risk budget", "Sizing, not Kelly"),
    ("entry", "Houston / Entry", "Taker"),
    ("exit", "Houston / Exit", "Closer"),
)
LIMITATIONS = [
    "Recorded stage summaries, not live agent events. Stage timing and confidence were not recorded.",
    "No frozen market-input archive in these reports. Current quotes and standalone scans are not joined to old cycles.",
    "Research is not linked to these decisions. Prior calibration, option-level edge and Kelly sizing are not implemented.",
    "Reservation risk pass is not human approval. Submission is not a verified fill. No approval or order controls exist here.",
]


def _unique_object(pairs):
    result = {}
    for key, value in pairs:
        if key in result:
            raise ValueError('Duplicate record key')
        result[key] = value
    return result


def obj(value):
    return value if isinstance(value, dict) else {}


def rows(value):
    return [r for r in value[:MAX_ROWS] if isinstance(r, dict)] if isinstance(value, list) else []


def text(value):
    if not isinstance(value, str):
        return None
    # Also redact common private identifiers embedded in otherwise allowed prose.
    value = re.sub(r"\b(?:account_id|authorization_id|client_order_id|broker_order_id|request_id|local_id)\s*[:=]\s*[^\s,;]+",
                   "[private reference redacted]", value, flags=re.I)
    return redact_secrets(value)[:400]


def enum(value, allowed):
    return value if isinstance(value, str) and value in allowed else None


def boolean(value):
    return value if type(value) is bool else None


def count(value):
    return value if type(value) is int and 0 <= value <= 1_000_000 else None


def number(value):
    if isinstance(value, bool) or not isinstance(value, (str, int, float, Decimal)):
        return None
    try:
        n = Decimal(str(value))
        return format(n, 'f') if n.is_finite() and abs(n) <= 1_000_000_000 and n.as_tuple().exponent >= -12 else None
    except (DecimalException, ValueError):
        return None


def stamp(value):
    try:
        d = datetime.fromisoformat(str(value).replace('Z', '+00:00'))
        return d.astimezone(timezone.utc).isoformat() if d.tzinfo else None
    except (ValueError, TypeError, OverflowError):
        return None


def submission(value):
    r = obj(value)
    return {"outcome": enum(r.get('outcome'), ('submitted', 'unknown', 'unplaced', 'not_submitted')),
            "broker_status": enum(r.get('broker_status'), ('new', 'accepted', 'pending_new', 'partially_filled', 'filled', 'canceled', 'expired', 'rejected', 'done_for_day')),
            "reason": text(r.get('reason'))}


def proposal(idea, contract=None):
    """Fixed fields only. A displayed estimate is never a rules verdict."""
    idea = obj(idea)
    legs = rows(idea.get('legs'))
    leg = legs[0] if len(legs) == 1 else {}
    q = count(idea.get('quantity'))
    debit, fees = number(idea.get('limit_debit')), number(idea.get('estimated_fees'))
    strategy = enum(idea.get('strategy'), ('long_call', 'long_put'))
    risk = None
    if strategy and len(legs) == 1 and q and debit is not None and fees is not None:
        if Decimal(debit) > 0 and Decimal(fees) >= 0:
            risk = number(Decimal(debit) * 100 * q + Decimal(fees))
    return {"symbol": text(idea.get('symbol')), "contract": text(contract),
            "strategy": strategy, "underlying_bias": enum(idea.get('direction'), ('long', 'short')),
            "right": enum(leg.get('right'), ('call', 'put')), "expiration": text(leg.get('expiration')),
            "strike": number(leg.get('strike')), "quantity": q, "multiplier": 100,
            "limit_premium_usd": debit, "estimated_fees_usd": fees, "estimated_max_risk_usd": risk,
            "underlying_entry": number(idea.get('entry_trigger')), "underlying_stop": number(idea.get('stop')),
            "underlying_target": number(idea.get('target')), "reasoning": text(idea.get('thesis')),
            "exit_plan": {"premium_stop_pct": number(obj(idea.get('exit_plan')).get('premium_stop_pct')),
                          "time_stop_dte": count(obj(idea.get('exit_plan')).get('time_stop_dte')),
                          "time_stop_sessions": count(obj(idea.get('exit_plan')).get('time_stop_sessions')),
                          "underlying_stop_rule": text(obj(idea.get('exit_plan')).get('underlying_stop_rule'))},
            "risk_basis": "Single-leg long premium × 100 × quantity + estimated fees; not underlying stop distance.",
            "confidence": None, "expected_option_edge": None}


def _journal(db, cycle_id, entry):
    # The existing controller's exact decision key, never 'latest for symbol'.
    symbol, playbook = entry.get('symbol'), entry.get('playbook')
    if (db is None or not re.fullmatch(r'[a-f0-9]{32}', cycle_id)
            or not isinstance(symbol, str) or not re.fullmatch(r'[A-Z]{1,6}', symbol)
            or not isinstance(playbook, str) or not re.fullmatch(r'[a-z_]{1,50}', playbook)):
        return None
    key = f'entry-{cycle_id}-{symbol}-{playbook}'
    r = db.execute("""SELECT substr(idea,1,16385) idea, status, contract, created_at
                       FROM order_intents WHERE decision_key=? AND kind='entry'""", (key,)).fetchone()
    if not r or len(r['idea']) > 16384:
        return None
    try:
        idea = json.loads(r['idea'], object_pairs_hook=_unique_object)
    except (ValueError, TypeError):
        return None
    if not isinstance(idea, dict) or idea.get('symbol') != symbol or idea.get('playbook') != playbook:
        return None
    return {"proposal": proposal(idea, r['contract']),
            "journal_status_now": enum(r['status'], ('reserved', 'claimed', 'rejected', 'expired', 'released',
                'resolved:filled', 'resolved:canceled', 'resolved:expired', 'resolved:rejected', 'resolved:done_for_day', 'resolved:unplaced', 'resolved:released_manual')),
            "journal_created_at": stamp(r['created_at']),
            "source": "orders.sqlite3 · exact controller decision-key match; status is current at read time"}


def project_cycle(raw, *, cycle_id, db=None):
    """Project one record independently. No joining of unrelated research/scans."""
    s = obj(raw.get('stages'))
    rec, scan, entry_stage = (obj(s.get(k)) for k in ('reconcile', 'scan', 'entries'))
    mode = enum(raw.get('control_mode'), ('DISABLED', 'EXITS_ONLY', 'ARMED_PAPER'))
    enabled = raw.get('enabled_playbooks')
    enabled = [text(p) for p in enabled[:MAX_ROWS]] if isinstance(enabled, list) and all(isinstance(p, str) for p in enabled) else None
    reconcile = {"ok": boolean(rec.get('ok')), "reasons": [text(r) for r in rec.get('reasons', [])[:MAX_ROWS] if isinstance(r, str)] if isinstance(rec.get('reasons'), list) else [],
                 "error": text(rec.get('error')), "open_positions": count(rec.get('open_positions')),
                 "pending": count(rec.get('pending')), "settled_cash": number(rec.get('settled_cash')),
                 "daily_loss": number(rec.get('daily_loss')), "breaker": boolean(rec.get('breaker'))}
    blocked = reconcile['ok'] is False or 'error' in rec or 'halt' in s
    skipped = text(entry_stage.get('skipped'))
    error = text(entry_stage.get('error'))
    entries = []
    for i, r in enumerate(rows(entry_stage.get('entries'))):
        item = {"id": f'{cycle_id}:entry:{i}', "symbol": text(r.get('symbol')), "playbook": text(r.get('playbook')),
                "risk_pass_at_reservation": boolean(r.get('reserved')), "reason": text(r.get('reason')),
                "claimed_at_cycle": boolean(r.get('claimed')), "claim_reason": text(r.get('claim_reason')),
                "submission": submission(r.get('submission')), "journal": None,
                "source": f'cycles.jsonl · stages.entries.entries[{i}]'}
        item['journal'] = _journal(db, cycle_id, r)
        entries.append(item)
    exits = []
    for i, r in enumerate(rows(s.get('exits'))):
        d = obj(r.get('decision'))
        exits.append({"id": f'{cycle_id}:exit:{i}', "contract": text(r.get('contract')), "note": text(r.get('note')),
                      "prepared": boolean(r.get('prepared')), "reason": text(r.get('reason')),
                      "submission": submission(r.get('submission')),
                      "decision": {"exit_reason": text(d.get('exit_reason')), "quantity": count(d.get('quantity')),
                                   "limit_price": number(d.get('limit_price'))},
                      "entry_link": None, "realized_pnl": None,
                      "source": f'cycles.jsonl · stages.exits[{i}]'})
    scan_view = {k: count(scan.get(k)) for k in ('shadow', 'proposals', 'skipped')}
    scan_view['refused'] = [{"symbol": text(r.get('symbol')), "reason": text(r.get('reason'))} for r in rows(scan.get('refused'))]
    if scan:
        scan_state, scan_note = 'recorded', 'Scan counts and gate refusals recorded. Signal inputs and shadow ideas are not archived here.'
    elif blocked or skipped or mode in ('DISABLED', 'EXITS_ONLY'):
        scan_state, scan_note = 'skipped', skipped or 'Cycle halted or entry mode does not permit scanning.'
    else:
        scan_state, scan_note = 'unknown', error or 'No scan stage recorded.'
    if error:
        scan_state = 'error'
    risk_state = 'rejected' if any(r['risk_pass_at_reservation'] is False for r in entries) else (
        'risk_pass' if entries and all(r['risk_pass_at_reservation'] is True for r in entries) else 'skipped' if skipped or blocked else 'unknown')
    entry_state = 'uncertain' if any(r['submission']['outcome'] == 'unknown' for r in entries) else (
        'submission_recorded' if any(r['submission']['outcome'] == 'submitted' for r in entries) else
        'rejected' if any(r['risk_pass_at_reservation'] is False
                          or (r['claimed_at_cycle'] is False and r['claim_reason'] and not r['claim_reason'].startswith('DRY_RUN:'))
                          or r['submission']['outcome'] == 'unplaced' for r in entries) else
        'dry_run' if entries and raw.get('submit') is False and any(r['risk_pass_at_reservation'] is True for r in entries) else
        'recorded' if entries else 'skipped' if skipped or blocked else 'error' if error else 'unknown')
    # The controller writes stages.exits (possibly []) after reconcile passes, but only evaluates exits
    # when there is inventory and the mode permits. Distinguish 'nothing to evaluate' from 'not evaluated'.
    exits_present = isinstance(s.get('exits'), list)
    if exits:
        exit_state, exit_note = 'recorded', 'Exit evaluations precede scanning. No entry linkage or realized P&L is inferred from a matching contract.'
    elif blocked:
        exit_state, exit_note = 'skipped', 'Cycle halted before exit evaluation.'
    elif exits_present and reconcile['open_positions'] == 0:
        exit_state, exit_note = 'recorded', 'Exit stage ran: no open option positions at this cycle, so nothing to evaluate.'
    elif exits_present and mode == 'DISABLED':
        exit_state, exit_note = 'skipped', 'DISABLED does not evaluate automatic exits; open positions were not managed this cycle.'
    else:
        exit_state, exit_note = 'unknown', 'No exit evaluations recorded; this does not prove there were no positions.'
    specs = {
        'spotter': (scan_state, scan_note, 'cycles.jsonl · stages.scan', scan_view),
        'prior': ('not_implemented', 'No calibrated prior. Saved research can be reviewed separately; it was not linked to this cycle.', None,
                  {"probability": None, "research_linked_to_cycle": False}),
        'edge': ('not_implemented', 'Option-level expectancy is not modeled. Underlying R is not an option edge estimate.', None,
                 {"expected_option_edge": None, "options_pnl_modelled": False}),
        'risk': (risk_state, 'Recorded reservation verdicts, not human approval or current permission. Cycle reconciliation is not the later reserve/claim input snapshot.',
                 'cycles.jsonl · stages.reconcile / stages.entries', {"cycle_reconciliation": reconcile, "entry_verdicts": entries}),
        'entry': (entry_state, skipped or error or 'Recorded entry attempts. Acknowledged submission is not a verified fill. Unknown outcomes require reconciliation, never retry.',
                  'cycles.jsonl · stages.entries', {"entries": entries, "skip_reason": skipped, "error": error}),
        'exit': (exit_state, exit_note, 'cycles.jsonl · stages.exits', {"exits": exits, "open_positions_at_reconcile": reconcile['open_positions']}),
    }
    agents = [{"id": key, "name": name, "role": role, "status": specs[key][0], "summary": specs[key][1],
               "source": specs[key][2], "outputs": specs[key][3], "confidence": None, "latency_ms": None,
               "model": None, "inputs": None} for key, name, role in ROLES]
    finished = stamp(raw.get('finished_at'))
    cycle_status = 'incomplete' if not finished else 'halted' if blocked else 'error' if error else entry_state
    if reconcile['ok'] is None or mode is None:
        cycle_status = 'unknown'
    if any(r['submission']['outcome'] == 'unknown' for r in exits):
        cycle_status = 'uncertain'
    flow = [
        {"name": "Reconcile", "status": 'blocked' if blocked else 'passed_at_cycle' if reconcile['ok'] is True else 'unknown'},
        {"name": "Existing exits", "status": exit_state},
        {"name": "Entry-mode gate", "status": 'skipped' if blocked else mode or 'unknown'},
        {"name": "Scan & candidate gates", "status": scan_state},
        {"name": "Reserve → optional claim / submit", "status": entry_state},
    ]
    activity = [{"label": step['name'], "status": step['status'], "recorded_at": finished,
                 "stage_timestamp": None, "detail": detail, "source": source}
                for step, detail, source in zip(flow, (reconcile, {"exits": exits}, {"mode": mode, "enabled_playbooks_at_cycle": enabled, "skip_reason": skipped}, scan_view, {"entries": entries, "error": error}),
                                               ('stages.reconcile', 'stages.exits', 'control_mode / stages.entries', 'stages.scan', 'stages.entries'))]
    return {"id": cycle_id, "started_at": stamp(raw.get('started_at')), "completed_at": finished,
            "status": cycle_status, "control_mode_at_cycle": mode, "enabled_playbooks_at_cycle": enabled,
            "submit_at_cycle": boolean(raw.get('submit')),
            "agents": agents, "flow": flow, "activity": activity, "entries": entries, "exits": exits,
            "market_snapshot": None, "human_approval": "Not available in Agent Desk",
            "limits_reference": {"capital_cap_usd": "2000", "maximum_entry_risk_usd": "100", "maximum_positions": 2,
                                 "cumulative_daily_loss_breaker_usd": "40", "note": "Current system limits, not a frozen historical configuration. No Kelly sizing."},
            "source": "cycles.jsonl; journal enrichment uses exact decision keys only",
            "details_truncated": any(isinstance(v, list) and len(v) > MAX_ROWS for v in (s.get('exits'), entry_stage.get('entries'), scan.get('refused'), rec.get('reasons')))}


def _load_cycles(path, warnings):
    if not path.exists():
        warnings.append('No cycles.jsonl record. Controller state is unknown.')
        return [], False
    try:
        with path.open('rb') as f:
            f.seek(0, 2)
            size = f.tell()
            start = max(0, size - MAX_BYTES)
            f.seek(start)
            data = f.read(MAX_BYTES)
        if start:
            data = data.partition(b'\n')[2]  # Never parse a cut-off first record.
        if data and not data.endswith(b'\n'):
            warnings.append('Trailing cycle record is incomplete; not displayed.')
            data = data.rpartition(b'\n')[0] + b'\n' if b'\n' in data else b''
        lines = data.splitlines()
        truncated = bool(start or len(lines) > MAX_CYCLES)
        result = []
        for line in reversed(lines[-MAX_CYCLES:]):
            try:
                if len(line) > MAX_LINE:
                    raise ValueError()
                value = json.loads(line.decode('utf-8'), object_pairs_hook=_unique_object)
                if not isinstance(value, dict) or not isinstance(value.get('stages'), dict):
                    raise ValueError()
                cid = value.get('cycle_id')
                if not isinstance(cid, str) or not re.fullmatch(r'[a-f0-9]{32}', cid):
                    cid = 'legacy-' + hashlib.sha256(line).hexdigest()[:24]
                    warnings.append('Legacy record has no valid controller cycle ID; journal linkage unavailable.')
                result.append((cid, value))
            except (ValueError, UnicodeError, RecursionError):
                warnings.append('Malformed or oversized cycle record skipped; history is incomplete.')
        duplicates = {cid for cid, n in Counter(cid for cid, _ in result).items() if n > 1}
        if duplicates:
            warnings.append('Duplicate cycle IDs omitted; no ambiguous history or journal linkage displayed.')
        return [(cid, r) for cid, r in result if cid not in duplicates], truncated
    except OSError:
        warnings.append('Cycle history unreadable. Controller state is unknown.')
        return [], False


def read_state(runtime: Path, *, now: datetime):
    warnings = []
    raw, truncated = _load_cycles(runtime / 'cycles.jsonl', warnings)
    db = None
    journal_path = runtime / 'orders.sqlite3'
    try:
        if raw:
            if journal_path.exists():
                db = sqlite3.connect(journal_path.resolve().as_uri() + '?mode=ro', uri=True, timeout=0.2)
                db.row_factory = sqlite3.Row
                db.execute('PRAGMA query_only=ON')
            else:
                warnings.append('Order journal unavailable; proposal economics and subsequent status are unknown.')
        try:
            cycles = [project_cycle(r, cycle_id=cid, db=db) for cid, r in raw]
        except (sqlite3.Error, TypeError, ValueError):
            warnings.append('Order journal unreadable; showing cycle evidence without journal enrichment.')
            cycles = [project_cycle(r, cycle_id=cid) for cid, r in raw]
    except sqlite3.Error:
        warnings.append('Order journal unreadable; showing cycle evidence without journal enrichment.')
        cycles = [project_cycle(r, cycle_id=cid) for cid, r in raw]
    finally:
        if db is not None:
            db.close()
    state = {"schema_version": 1, "cycles": cycles, "warnings": list(dict.fromkeys(warnings)),
             "history_truncated": truncated, "history_limit": MAX_CYCLES, "limitations": LIMITATIONS}
    state['cursor'] = hashlib.sha256(json.dumps(state, sort_keys=True, allow_nan=False).encode()).hexdigest()
    state['read_at'] = now.astimezone(timezone.utc).isoformat()
    return state


def route(path):
    """Recognize only bounded read routes; no arbitrary filenames or write verbs."""
    if len(path) > 240:
        return None
    u = urlsplit(path)
    if u.fragment or u.scheme or u.netloc:
        return None
    if u.path == '/api/agent-desk/state':
        q = parse_qs(u.query, keep_blank_values=True)
        if not q:
            return ('state', None)
        if set(q) == {'after'} and len(q['after']) == 1 and re.fullmatch(r'[a-f0-9]{64}', q['after'][0]):
            return ('state', q['after'][0])
    if not u.query:
        if u.path == '/api/agent-desk/cycles':
            return ('cycles', None)
        prefix = '/api/agent-desk/cycles/'
        if u.path.startswith(prefix) and ID.fullmatch(u.path[len(prefix):]):
            return ('cycle', u.path[len(prefix):])
    return None


def response(state, kind, value):
    if kind == 'state':
        return {"unchanged": True, "cursor": state['cursor'], "read_at": state['read_at']} if value == state['cursor'] else state
    if kind == 'cycles':
        return {"cycles": [{k: c[k] for k in ('id', 'started_at', 'completed_at', 'status', 'control_mode_at_cycle')} for c in state['cycles']],
                "cursor": state['cursor'], "history_truncated": state['history_truncated'], "warnings": state['warnings']}
    return next((c for c in state['cycles'] if c['id'] == value), None)

File: src/alpaca_agents/assets/agent_desk.js
/* Agent Desk is a passive reader. All rendering uses DOM nodes and text.
   No order, approval, controller-run, market-data, or model endpoint is called. */
function initializeAgentDesk({state, request, live, canRead}) {
  const panel = document.getElementById('desk-panel');
  if (!panel || !state) return;
  const select = document.getElementById('desk-cycle');
  const follow = document.getElementById('desk-follow');
  const refreshButton = document.getElementById('desk-refresh');
  const connection = document.getElementById('desk-connection');
  const inspector = document.getElementById('desk-inspector');
  const cards = [...document.querySelectorAll('[data-desk-agent]')];
  const replayButton = document.getElementById('desk-replay');
  const swarm = document.querySelector('.desk-swarm');
  const core = document.querySelector('.desk-core');
  const links = [...document.querySelectorAll('.desk-connectors path')];
  let following = true, selected = state.cycles[0] || null, role = 'spotter';
  let reading = false, disconnected = false, replayToken = 0;
  // Mirror of TONES in agent_desk_view.py. Unknown statuses are amber, never green.
  const TONES = {passed_at_cycle: 'pass', recorded: 'pass', risk_pass: 'pass', submission_recorded: 'pass',
                 dry_run: 'pass', ARMED_PAPER: 'pass', EXITS_ONLY: 'warn',
                 blocked: 'stop', rejected: 'stop', error: 'stop', uncertain: 'stop', halted: 'stop'};
  const tone = status => TONES[status] || 'warn';
  const reducedMotion = matchMedia('(prefers-reduced-motion: reduce)');
  const node = (tag, text, cls) => {
    const element = document.createElement(tag);
    if (text !== undefined) element.textContent = text;
    if (cls) element.className = cls;
    return element;
  };
  const shown = value => value === null || value === undefined ? 'Unknown' : typeof value === 'boolean' ? (value ? 'Yes (recorded)' : 'No (recorded)') : String(value);
  const human = value => shown(value).replace(/_/g, ' ');
  function fields(items) {
    const dl = node('dl');
    items.forEach(([label, value]) => dl.append(node('dt', label), node('dd', shown(value))));
    return dl;
  }
  function json(value) { return node('pre', JSON.stringify(value, null, 2)); }
  function freshness() {
    const when = Date.parse(selected && selected.completed_at);
    const delta = Date.now() - when;
    const age = !Number.isFinite(when) ? 'Completion time unknown' : delta < 0 ? 'Future timestamp — check clock' :
      delta > 60000 ? `Historical / stale cycle · ${Math.floor(delta / 60000)} min since completion` : 'Recently recorded cycle, not proof of a running controller';
    const retained = selected && !state.cycles.some(c => c.id === selected.id);
    document.getElementById('desk-freshness').textContent = `${age}. Last read: ${shown(state.read_at)}. ${retained ? 'Pinned snapshot is outside the current history window; its journal status is not refreshed.' : 'Journal status, when linked, is current at that read; cycle decisions remain historical.'}`;
  }
  function inspect(agent, focus = false) {
    inspector.replaceChildren();
    cards.forEach(card => card.setAttribute('aria-pressed', String(agent && card.dataset.deskAgent === agent.id)));
    if (!agent) {
      inspector.append(node('h3', 'No recorded cycle'), node('p', 'Unknown is not idle. Viewing the desk does not start a cycle.', 'dim'));
      return;
    }
    inspector.append(node('h3', agent.name), node('p', agent.summary), fields([
      ['Recorded state', human(agent.status)], ['Confidence', agent.confidence], ['Latency (ms)', agent.latency_ms],
      ['Model', agent.model], ['Input snapshot', agent.inputs === null ? 'Not captured in these reports' : shown(agent.inputs)],
      ['Source', agent.source || 'Not implemented / no cycle evidence'], ['Cycle', selected.id]
    ]), json(agent.outputs));
    if (focus) inspector.focus({preventScroll: true});
  }
  function drawProposals() {
    const box = document.getElementById('desk-proposals');
    box.replaceChildren(node('h3', 'Entry proposals & recorded risk'));
    if (!selected || !selected.entries.length) {
      box.append(node('p', 'No evaluated entry records in this cycle. This does not imply no opportunities or a completed scan.', 'desk-empty'));
      return;
    }
    selected.entries.forEach(entry => {
      const detail = node('details', undefined, 'desk-proposal');
      const heading = node('summary');
      heading.append(node('strong', `${shown(entry.symbol)} · ${shown(entry.playbook)}`),
        node('span', entry.risk_pass_at_reservation === true ? 'Risk passed at reservation · not human approval' : entry.risk_pass_at_reservation === false ? 'Rejected at reservation' : 'Risk verdict unknown'));
      detail.append(heading, fields([
        ['Reason', entry.reason], ['Claimed at cycle', entry.claimed_at_cycle], ['Claim reason', entry.claim_reason],
        ['Submission outcome at cycle', entry.submission.outcome], ['Broker status at submission', entry.submission.broker_status],
        ['Journal status at last read', entry.journal && entry.journal.journal_status_now]
      ]));
      const p = entry.journal && entry.journal.proposal;
      if (p) {
        detail.append(fields([
          ['Contract', p.contract], ['Strategy', p.strategy], ['Underlying bias (not order side)', p.underlying_bias],
          ['Call / put', p.right], ['Expiry', p.expiration], ['Strike', p.strike], ['Contracts', p.quantity], ['Multiplier', p.multiplier],
          ['Limit premium / share (USD)', p.limit_premium_usd], ['Estimated fees (USD)', p.estimated_fees_usd],
          ['Estimated maximum entry risk (USD)', p.estimated_max_risk_usd], ['Underlying entry', p.underlying_entry],
          ['Underlying stop', p.underlying_stop], ['Underlying target', p.underlying_target],
          ['Saved premium stop (%)', p.exit_plan.premium_stop_pct], ['Saved time stop (DTE)', p.exit_plan.time_stop_dte],
          ['Saved time stop (sessions)', p.exit_plan.time_stop_sessions], ['Saved underlying exit rule', p.exit_plan.underlying_stop_rule],
          ['Recorded reasoning', p.reasoning]
        ]), node('p', p.risk_basis, 'desk-source'), node('p', entry.journal.source, 'desk-source'));
      } else detail.append(node('p', 'Proposal economics unknown: no exact journal match. Current scans or quotes are not substituted.', 'desk-empty'));
      detail.append(node('p', `${entry.source} · ${entry.id}`, 'desk-source'));
      box.append(detail);
    });
    box.append(node('p', 'Claimed / unknown submission requires reconciliation, never a retry. A submitted order is not a verified fill. No approval or submission controls.', 'desk-source'));
  }
  function drawActivity() {
    const box = document.getElementById('desk-activity');
    box.replaceChildren(node('h3', 'Recorded activity'));
    box.append(node('p', 'Stage summaries in controller order. Individual event times were not captured; these are not live agent messages.', 'desk-source'));
    if (!selected) return;
    selected.activity.forEach(item => {
      const detail = node('details', undefined, 'desk-activity-item');
      const heading = node('summary');
      heading.append(node('strong', item.label), node('span', human(item.status)));
      detail.append(heading, node('p', `Report completed: ${shown(item.recorded_at)} · source: cycles.jsonl / ${item.source}`, 'desk-source'), json(item.detail));
      box.append(detail);
    });
  }
  function draw() {
    const list = state.cycles.slice();
    if (selected && !list.some(c => c.id === selected.id)) list.push(selected);
    select.replaceChildren();
    if (!list.length) select.append(node('option', 'No recorded cycles'));
    list.forEach(c => {
      const option = node('option', `${shown(c.completed_at)} · ${human(c.status)} · ${c.id.slice(0, 12)}`);
      option.value = c.id;
      select.append(option);
    });
    select.disabled = !list.length;
    if (selected) select.value = selected.id;
    follow.disabled = !state.cycles.length;
    follow.setAttribute('aria-pressed', String(following));
    follow.textContent = following ? 'Following latest' : 'Follow latest';
    document.getElementById('desk-cycle-title').textContent = selected ? `${selected.id} · ${shown(selected.control_mode_at_cycle)} · submission setting: ${shown(selected.submit_at_cycle)}` : 'No recorded cycle';
    const flow = document.querySelector('.desk-flow');
    flow.replaceChildren();
    if (selected) selected.flow.forEach(step => {
      const li = node('li'); li.dataset.tone = tone(step.status);
      li.append(node('strong', step.name), node('span', human(step.status))); flow.append(li);
    });
    else flow.append(node('li', 'No recorded flow. Unknown is not idle.'));
    cards.forEach(card => {
      const agent = selected && selected.agents.find(a => a.id === card.dataset.deskAgent);
      card.disabled = !agent;
      card.dataset.status = agent ? agent.status : 'unknown';
      card.dataset.tone = tone(card.dataset.status);
      card.querySelector('.desk-state').textContent = agent ? human(agent.status) : 'Unknown';
    });
    core.dataset.tone = selected ? tone(selected.status) : 'warn';
    document.querySelector('.desk-core strong').textContent = selected ? human(selected.status) : 'Unknown';
    replayButton.disabled = !selected;
    replay();
    inspect(selected && selected.agents.find(a => a.id === role));
    drawProposals(); drawActivity();
    document.getElementById('desk-evidence').textContent = selected ? JSON.stringify(selected, null, 2) : 'No saved cycle. This page does not start one.';
    const warnings = document.getElementById('desk-warnings');
    warnings.replaceChildren(...state.warnings.map(w => node('li', w)));
    if (selected && selected.details_truncated) warnings.append(node('li', 'Cycle details exceed the display bound; additional rows are omitted.'));
    document.getElementById('desk-history-note').textContent = `Up to ${state.history_limit} saved cycles from a bounded file tail. ${state.history_truncated ? 'Older history omitted.' : 'History may be incomplete.'} Research and fills are not joined by proximity.`;
    freshness();
  }
  // Replays the SAVED decision path: each stage lights in controller order and the signal only
  // advances past a stage the record shows as passed. It stops, amber or red, where the cycle
  // ended. Pure presentation of stored statuses; nothing is polled, predicted or executed.
  function replay() {
    const token = ++replayToken;
    const steps = [...document.querySelectorAll('.desk-flow li')];
    steps.forEach(li => li.classList.remove('lit', 'signal'));
    cards.forEach(c => c.classList.remove('lit'));
    core.classList.remove('lit');
    links.forEach(p => p.classList.remove('lit'));
    swarm.classList.remove('flowing');
    if (!selected) return;
    const instant = reducedMotion.matches;
    const at = (ms, fn) => instant ? fn() : setTimeout(() => { if (token === replayToken) fn(); }, ms);
    let delay = 0, open = true;
    steps.forEach((li, i) => {
      if (!open) return;
      at(delay, () => li.classList.add('lit'));
      if (li.dataset.tone === 'pass' && i < steps.length - 1) {
        at(delay + 250, () => li.classList.add('signal'));
        delay += 480;
      } else open = false;
    });
    delay += 350;
    const light = id => {
      const card = cards.find(c => c.dataset.deskAgent === id);
      const link = links.find(p => p.dataset.link === id);
      if (card) card.classList.add('lit');
      if (link && card) { link.dataset.tone = card.dataset.tone; link.classList.add('lit'); }
    };
    at(delay, () => swarm.classList.add('flowing'));
    ['spotter', 'prior', 'edge'].forEach((id, i) => at(delay + i * 160, () => light(id)));
    delay += 620;
    at(delay, () => core.classList.add('lit'));
    delay += 320;
    ['risk', 'entry', 'exit'].forEach((id, i) => at(delay + i * 160, () => light(id)));
    at(delay + 1800, () => swarm.classList.remove('flowing'));
  }
  async function refresh() {
    if (!live || reading || !panel.open || document.hidden || !canRead()) return;
    reading = true;
    refreshButton.disabled = true;
    connection.textContent = 'Reading local records… may wait behind controller work. Existing evidence is not a live status.';
    try {
      const next = await request('/api/agent-desk/state' + (state.cursor ? '?after=' + state.cursor : ''));
      if (next.unchanged === true && next.cursor === state.cursor) state.read_at = next.read_at;
      else {
        if (next.schema_version !== 1 || !Array.isArray(next.cycles) || !Array.isArray(next.warnings)) throw new Error('Invalid desk state');
        state = next;
        selected = following ? state.cycles[0] || null : state.cycles.find(c => selected && c.id === selected.id) || selected;
        draw();
      }
      disconnected = false;
      connection.textContent = 'Connected to local records · polls every 15s while open · not a live stage stream.';
    } catch {
      disconnected = true;
      connection.textContent = 'Disconnected / read refused — current state unknown. Showing retained evidence only; no cycle or order retry.';
    } finally {
      reading = false;
      refreshButton.disabled = false;
      freshness();
    }
  }
  cards.forEach(card => card.addEventListener('click', () => {
    role = card.dataset.deskAgent;
    inspect(selected && selected.agents.find(a => a.id === role), true);
  }));
  select.addEventListener('change', () => {
    const found = state.cycles.find(c => c.id === select.value);
    if (found) selected = found;
    following = false;
    draw();
  });
  follow.addEventListener('click', () => { following = true; selected = state.cycles[0] || null; draw(); });
  replayButton.addEventListener('click', replay);
  refreshButton.disabled = !live;
  refreshButton.hidden = !live;
  refreshButton.addEventListener('click', refresh);
  panel.addEventListener('toggle', () => { if (panel.open) refresh(); });
  connection.textContent = live ? 'Local recorded state · open the desk to refresh. Not a live stage stream.' : 'Offline snapshot · no network · rebuild to update.';
  draw();
  setInterval(() => {
    freshness();
    // A failed read needs an explicit refresh/reopen, not an automatic retry loop.
    if (!disconnected) refresh();
  }, 15000);
}
